scheduler/system_util.go

```diff
@@ -8,24 +8,26 @@ package scheduler
 
 import (
 	"fmt"
+	"sort"
 	"time"
 
 	"github.com/hashicorp/nomad/nomad/structs"
 )
 
-// materializeSystemTaskGroups is used to materialize all the task groups
-// a system or sysbatch job requires.
+// materializeSystemTaskGroups is used to materialize all the task groups a
+// system or sysbatch job requires. Returns a map of allocation names to task
+// groups.
 func materializeSystemTaskGroups(job *structs.Job) map[string]*structs.TaskGroup {
 	out := make(map[string]*structs.TaskGroup)
 	if job.Stopped() {
 		return out
 	}
 
 	for _, tg := range job.TaskGroups {
-		for i := 0; i < tg.Count; i++ {
-			name := fmt.Sprintf("%s.%s[%d]", job.Name, tg.Name, i)
-			out[name] = tg
-		}
+		// note that all system/sysbatch allocs for the same task group share
+		// the same alloc index 0
+		name := fmt.Sprintf("%s.%s[0]", job.Name, tg.Name)
+		out[name] = tg
```

Comment:
> I believe tg.Count is always 1 for system jobs, but not for sysbatch?

`tg.Count` is ignored for system jobs, because we always run 1 per node that meets the requirements. You can omit the field, which means that most of the time the value will be 0.

I suspect allowing `tg.Count > 1` for `sysbatch` jobs is just a bug in validation. If you actually run a `sysbatch` job with `count=3` with 2 eligible clients, you get 1 per node. The same behavior we get with `system` jobs, which is what I'd expect.

Reply:
> If you actually run a sysbatch job with count=3 with 2 eligible clients, you get 1 per node.

Did you test using a binary built with this branch? From `main` and 1.4.3 I get 4 allocations if you run a job like this in a cluster with 2 clients.

```hcl
job "example" {
  type        = "sysbatch"
  datacenters = ["dc1"]

  group "echo" {
    count = 2

    task "echo" {
      driver = "raw_exec"

      config {
        command = "echo"
        args    = ["hi"]
      }
    }
  }
}
```

```console
$ nomad job status example
ID            = example
Name          = example
Submit Date   = 2023-02-09T14:04:09-05:00
Type          = sysbatch
Priority      = 50
Datacenters   = dc1
Namespace     = default
Status        = dead
Periodic      = false
Parameterized = false

Summary
Task Group  Queued  Starting  Running  Failed  Complete  Lost  Unknown
echo        0       0         0        0       4         0     0

Allocations
ID        Node ID   Task Group  Version  Desired  Status    Created  Modified
65b2be15  cd72b386  echo        0        run      complete  8s ago   7s ago
7f67849c  327ab09f  echo        0        run      complete  8s ago   7s ago
93a4a3b2  327ab09f  echo        0        run      complete  8s ago   7s ago
f244a8ff  cd72b386  echo        0        run      complete  8s ago   7s ago
```

Running a binary from this branch gives me only 2 allocs.

I don't have any preference on either allowing or preventing count > 1 for sysbatch, but if we add a validation to prevent it it would be a breaking change.

Reply:
Shoot, I was certain I'd run it against `main` but maybe I forgot to rebuild after I checked out `main`. :blush: You're right that this is how sysbatch works, which means this entire patch is broken for sysbatch as-is, because we're enforcing a single ignored/updated/reconnected alloc.

I'm kinda of the opinion that this is a sysbatch bug, but the [RFC](https://docs.google.com/document/d/1FlYV71mRmC_qTHfXggsXD7ONICtyFqe5_yQDqHNsuU8/edit) doesn't say either way. Lemme bring this up with the team.

Reply:
Oof, just realized there's a fairly serious bug here if the system job has more than one task group